c++/src/kj/refcount.h

```diff
@@ -115,6 +142,106 @@ Own<T> Refcounted::addRefInternal(T* object) {
   return Own<T>(object, *refcounted);
 }
 
+template <typename T>
+Rc<T> Refcounted::addRcRefInternal(T* object) {
+  Refcounted* refcounted = object;
+  ++refcounted->refcount;
+  return Rc<T>(object);
+}
+
+template<typename T>
+class Rc: private Own<T> {
+  // Smart pointer for reference counted objects. 
+  //
+  // There are only two ways to obtain new Rc instances:
+  // - use kj::rc<T>(...) function to create new T.
+  // - use kj::Rc::addRef() and the existing Rc instance.
+  // - use EnableAddRefToThis to allow T instance to add new references to itself.
+  //
+  // kj::Rc<T> can be used everywhere Own<T> is expected.
+  //
+  // The suggested usage patterns are:
+  // - return kj::Rc as value from factory functions: 
+  //     kj::Rc<MyService> createMyService();
+  // - pass kj::Rc as rvalue to functions that need to extend T's lifetime:
+  //     void setMyService(kj::Rc<MyService>&& service)
+  // - store kj::Rc as data member:
+  //     struct MyComputation { kj::Rc<MyService> service; };
+  //
+
+public:
+  KJ_DISALLOW_COPY(Rc);
+  Rc() { }
+
+  inline Rc(Rc&& other) noexcept: Own<T>(kj::mv(other)) { }
```

Comment:
Nitpick, I think your move constructor and assignment operator can just be defaulted rather than fully defined.